README.md
# options-strategy-visualizer
* Small tool that plots the total payoff of a strategy combining financial call/put options and/or the underlying
* [![Build Status](https://secure.travis-ci.org/shadiakiki1986/options-strategy-visualizer.png)](http://travis-ci.org/shadiakiki1986/options-strategy-visualizer)

# Motivation
The reason I wrote this tool is that after some [research](http://lmgtfy.com/?q=options+strategy+visualize) for an online tool that visualizes option strategies, I didn't find anything useful.

This tool is hosted [here](http://genesis.akikieng.com/options-strategy-visualizer)

Its calculations are based on CBOE's [rulebook](http://www.cboe.com/micro/margin/strategy.aspx) (p 18)

Here is a list of pages that I checked out before I decided to do this myself
* https://discoveroptions.com/mixed/content/education/articles/riskgraphs.html 
* http://www.intrepid.com/robertl/strategies1.html
* http://www.mathworks.com/matlabcentral/fileexchange/17411-visualize-payoffs-of-an-option-strategy/content/butterfly.m
* http://www.optiontradingtips.com/options101/payoff-diagrams.html


# Logo
I chose the letter '''V''' to be the logo for this tool.

It represents 
* the first letter of the word '''visualizer''' 
* the payoff of a long [straddle](https://en.wikipedia.org/wiki/Straddle) position
* one of my [favorite movies](https://en.wikipedia.org/wiki/V_for_Vendetta)

# License
Licensed under [WTFPL](http://www.wtfpl.net/)


CHANGELOG.md
2015-06-15
* added php api backend with hook to calculate margin (along with tests of course)
* api: Tmt to be per option
* javascript: add physical time
* api: round 2nd decimal
* index: add link to github
* api: add security (needed to refactor c to C and p to P and add S)
* api: add offset from market value
* CI with travis-ci.org

2015-06-12
* can set time to maturity for each instrument (allows calendar spreads)
* higher resolution of sliders
* added presets for straddles and spreads (non-calendar + calendar)

2015-06-..
* first import


src/BlackScholes.php
<?php
# CND and value functions from: http://cseweb.ucsd.edu/~goguen/courses/130/SayBlackScholes.html

class BlackScholes {

var $call_put_flag, $X, $r, $v, $T, $V;

function __construct($call_put_flag, $X, $r, $v, $T, $V=0) {
// call_put_flag: 'c' or anything else for put
// This T is the maturity date (compare to t in value function)
// X: strike
// r: interest rates
// v: volatility
// V: market value

	if(!in_array($call_put_flag,array("C","P","S"))) throw new Exception("Unsupported call_put_flag=$call_put_flag");
	$this->call_put_flag=$call_put_flag;
	$this->X=$X;
	$this->r=$r;
	$this->v=$v;
	$this->T=$T;
	$this->V=$V;
}

	function CND ($x) {
		$Pi = 3.141592653589793238;
		$a1 = 0.319381530;
		$a2 = -0.356563782;
		$a3 = 1.781477937;
		$a4 = -1.821255978;
		$a5 = 1.330274429;
		$L = abs($x);
		$k = 1 / ( 1 + 0.2316419 * $L);
		$p = 1 - 1 /  pow(2 * $Pi, 0.5) * exp( -pow($L, 2) / 2 ) * ($a1 * $k + $a2 * pow($k, 2)
		     + $a3 * pow($k, 3) + $a4 * pow($k, 4) + $a5 * pow($k, 5) );
		if ($x >= 0) { 
			return $p;
		} else {
			return 1-$p;
		}
	}

	function value ($S, $t) {
	// t: physical time
		if($this->call_put_flag=="S") {
			$o=array();
			foreach($S as $Si) $o[(string)$Si]=$Si;
			return $o;
		}
		if(is_array($S)) {
			$o=array();
			foreach($S as $Si) $o[(string)$Si]=$this->value($Si,$t);
			return $o;
		}
		if(is_array($t)) {
			$o=array();
			foreach($t as $ti) $o[(string)$ti]=$this->value($S,$ti);
			return $o;
		}

		$Tmt=$this->T-$t;
		if($Tmt==0) return $this->value0($S);
		if($Tmt<0) return 0; // expired

		$d1 = ( log($S / $this->X) + ($this->r + pow($this->v, 2) / 2) * $Tmt ) / ( $this->v * pow($Tmt, 0.5) );
		$d2 = $d1 - $this->v * pow($Tmt, 0.5);
		if ($this->call_put_flag == 'C') {
			return $S * $this->CND($d1) - $this->X * exp( -$this->r * $Tmt ) * $this->CND($d2) - $this->V;
		} else { 
			return $this->X * exp( -$this->r * $Tmt ) * $this->CND(-$d2) - $S * $this->CND(-$d1) - $this->V;
		}
	}

	function value0($S) {
		if ($this->call_put_flag == 'C') {
			return max($S - $this->X, 0) - $this->V;
		} else { 
			return max($this->X - $S, 0) - $this->V;
		}
	}

function id() {
	return "".$this->call_put_flag."_".$this->X."_".$this->r."_".$this->v."_".$this->T; // not appending V here because V is the market value, and it should be the same for both options if they are similar
}

} // end class


src/OptionsStrategy.php
<?php

require_once dirname(__FILE__).'/../root.php';
require_once ROOT.'/src/BlackScholes.php';

class OptionsStrategy {

var $portfolio;

function __construct($portfolio) {
// array of BlackScholes objects

	if(!is_array($portfolio)) throw new Exception("Portfolio should be an array");
	if(count($portfolio)==0) throw new Exception("Portfolio is empty");
	array_map(function($p) {
		if(!array_key_exists("o",$p)) throw new Exception("Portfolio missing o key");
		if(!array_key_exists("q",$p)) throw new Exception("Portfolio missing q key");
		if(!is_int($p["q"])) throw new Exception("quantity should be numeric integer");
		if(! $p["o"] instanceof BlackScholes ) throw new Exception("Portfolio should be array of BlackScholes objects");
	}, $portfolio);

	// test no duplicates
	$ids=array_map(function($p) { return $p["o"]->id(); },$portfolio);
	if(count($ids)!=count(array_unique($ids))) throw new Exception("Duplicate options in portfolio detected");

	// portfolio being keyed by portfolio ID and stored in class member
	$this->portfolio=array();
	foreach($portfolio as $p) $this->portfolio[$p["o"]->id()]=$p;
}

function loss($S,$t,$mult=1) {
// $S: underlying value
// $t: physical time
// $mult: factor with which to multiply the loss... useful for this->margin

	if(!is_numeric($mult)) throw new Exception("mult should be numeric");
	return $this->loss2($this->loss1($S,$t),$mult,$S,$t);
}

function loss1($S,$t) {
// sub-function of loss
	$vals=array();
	foreach($this->portfolio as $id=>$p) {
		$v=$p["o"]->value($S,$t);
		if(is_array($v)) {
			$q=$p["q"];
			// 1st dimension: S
			$v=array_map(function($vi) use($q) {
				if(is_array($vi)) {
					// 2nd dimension: t
					return array_map(function($vii) use($q) { return $vii*$q; },$vi);
				} else {
					return $vi*$q;
				}
			}, $v); 
		} else {
			$v=$v*$p["q"];
		}

		// swap dimensions of options with (S,t)
		if(is_array($v)) {
			foreach($v as $ki=>$vi) {
				if(!array_key_exists($ki,$vals)) $vals[$ki]=array();
				if(is_array($vi)) {
					foreach($vi as $kii=>$vii) {
						if(!array_key_exists($kii,$vals[$ki])) $vals[$ki][$kii]=array();
						$vals[$ki][$kii][$id]=$vii;
					}
				} else {
					$vals[$ki][$id]=$vi;
				}
			}
		} else {
			$vals[$id]=$v;
		}
	}

	return $vals;
}

function loss2($vals,$mult,$S,$t) {
// sub-function of loss
	if(!is_array($S)&&!is_array($t)) {
		$vals = array_sum($vals);
		$vals = min($vals,0)*$mult;
	} else {
		if(is_array($S)&&is_array($t)) {
			foreach($vals as $ki=>$vi) {
				foreach($vi as $kii=>$vii) {
					$vii = array_sum($vii);
					$vii = min($vii,0)*$mult;
					$vals[$ki][$kii] = $vii;
				}
			}
		} else {
			if(!is_array($S)||!is_array($t)) {
				foreach($vals as $ki=>$vi) {
					$vi = array_sum($vi);
					$vi = min($vi,0)*$mult;
					$vals[$ki] = $vi;
				}
			} else {
				throw new Exception("WTF");
			}
		}
	}
	return $vals;
}

function margin($S,$t) {
	$m=$this->loss($S,$t,-1);

	if(!is_array($S)&&!is_array($t)) {
		if(count($this->portfolio)==1) return $m; // this would be numeric in this case
		return $m;
	} else {
		if(is_array($S)&&is_array($t)) {
			return max(max($m));
		} else {
			if(!is_array($S)||!is_array($t)) {
				return max($m);
			} else {
				throw new Exception("WTF");
			}
		}
	}
}

} // end class


www/api/optionsStrategyMargin.php
<?php

header("Access-Control-Allow-Origin: *");

/*
 Margin required on an options strategy

 USAGE
	CLI	php optionsStrategyMargin.php 100 0 '[{"Q": 1,"O":"C","K": 70,"r":0,"v":0.25,"T":0.2},{"Q":-1,"O":"C","K": 80,"r":0,"v":0.25,"T":0.2}]'
		php optionsStrategyMargin.php 100 0 '[{"Q":-1,"O":"C","K": 70,"r":0,"v":0.25,"T":0.2},{"Q": 1,"O":"C","K": 80,"r":0,"v":0.25,"T":0.2}]'
		php optionsStrategyMargin.php 125 0 '[{"Q":-1,"O":"C","K":125,"r":0,"v":0.25,"T":0.2},{"Q": 1,"O":"C","K":100,"r":0,"v":0.25,"T":0.2}]'

	AJAX
		 $.ajax({
		    url:'http://shadi.ly/zboota-server/api/optionsStrategyMargin.php?S=100&t=0.2&port=[{"Q":1,"O":"C","K":70,"r":0,"v":0.25},{"Q":-1,"O":"C","K":80,"r":0,"v":0.25}]',
		    success: function (data) {
			console.log(data);
		    },
		    error: function (jqXHR, ts, et) {
			console.log("error", ts, et);
		    }
		 });
*/

if(isset($argc)) {
	if($argc<3) throw new Exception("Usage: php optionsStrategyMargin.php underlying Tmt portfolio");
	$S=$argv[1];
	$t=$argv[2];
	$port=$argv[3];
} else {
	if(!array_key_exists("S",$_GET)||!array_key_exists("t",$_GET)||!array_key_exists("port",$_GET)) throw new Exception("Wrong usage of GET");
	$S=$_GET["S"];
	$t=$_GET["t"];
	$port=$_GET["port"];
}

require_once dirname(__FILE__).'/../../root.php';
require_once ROOT.'/src/OptionsStrategy.php';
require_once ROOT.'/src/BlackScholes.php';

if($port=="") throw new Exception("Please enter your portfolio.\n");
if($S=="") throw new Exception("Please enter underlying.\n");
if($t=="") throw new Exception("Please enter maturity.\n");

// read from json
$port=json_decode($port,true);
// convert to objects
$S=(float)$S;
$t=(float)$t;
$port2=array();
foreach($port as $pi) array_push($port2,array("q"=>$pi["Q"],"o"=>new BlackScholes($pi["O"],$pi["K"],$pi["r"],$pi["v"],$pi["T"],$pi["V"])));

// build vector of S around underlying
$Sx=range(0,2*$S,1);

try {
	// calculate
	$zc=new OptionsStrategy($port2);
	echo round($zc->margin($Sx,$t)*100)/100;
} catch (Exception $e) {
        echo $e->getMessage();
        return;
}



www/js/main.js
myJsonParse=function(aaa) {
	var oo=JSON.parse(aaa);
	oo.Q=parseInt(oo.Q);
	oo.K=parseInt(oo.K);
	oo.V=Number(oo.V);
	oo.St=parseInt(oo.St);
	if(!oo.hasOwnProperty("Tmt")) alert("No T-t set"); else oo.Tmt=Number(oo.Tmt);
	return oo;
};

// http://stackoverflow.com/questions/5259421/cumulative-distribution-function-in-javascript
function normalcdf(mean, sigma, to) 
{
    var z = (to-mean)/Math.sqrt(2*sigma*sigma);
    var t = 1/(1+0.3275911*Math.abs(z));
    var a1 =  0.254829592;
    var a2 = -0.284496736;
    var a3 =  1.421413741;
    var a4 = -1.453152027;
    var a5 =  1.061405429;
    var erf = 1-(((((a5*t + a4)*t) + a3)*t + a2)*t + a1)*t*Math.exp(-z*z);
    var sign = 1;
    if(z < 0)
    {
        sign = -1;
    }
    return (1/2)*(1+sign*erf);
}

function pnorm(to) { return normalcdf(0,1,to); }

// calculate option P&L (at maturity)
function mygv2(id,St) {
	var oo=JSON.parse($(id).prop("value"));
	var s=oo.P=="L"?1:(oo.P=="S"?-1:alert("wtf"));
	var vc=Math.max(0,St-oo.K);
	var vp=Math.max(0,oo.K-St);
	var vs=St-oo.K;
	var v=oo.O=="C"?vc:(oo.O=="P"?vp:(oo.O=="S"?St:alert("wtf2"))); // in case of stock, K is for the purchase price of the stock
	var p=oo.O=="C"||oo.O=="P"?oo.V:0;
	return(oo.Q*s*(v-p));
}

// http://en.wikipedia.org/wiki/Black%E2%80%93Scholes_model#Black-Scholes_formula
// assuming interest rates = 0 %
function calcd1(sig,Tmt,S,K,r) { return 1/sig/Math.sqrt(Tmt)*(Math.log(S/K)+(r+Math.pow(sig,2)/2)*Tmt); }
function calcd2(sig,Tmt,S,K,r) { return 1/sig/Math.sqrt(Tmt)*(Math.log(S/K)+(r-Math.pow(sig,2)/2)*Tmt); }

// calculate option price curve versus underlying (using Black-Scholes and implied vol=25% and interest rates=0%)
// http://en.wikipedia.org/wiki/Black%E2%80%93Scholes_model#Black-Scholes_formula
// Tmt is the number of years to maturity, e.g. 14 months = 14/12
// r=0 makes futures options similar to options on securities(the formula is modified for futures options   http://web.am.qub.ac.uk/users/m.s.kim/chap7.pdf  )
// Interest rates could be obtained from the table built-in the Eurex program: Settings / Interest Rates (minimize the subwindow to see the menu)
// but the effect is minimal really
// St is the underlying
function mygv(id,St) {
	var oo=myJsonParse($(id).prop("value"));
	if(oo.Tmt==0) return mygv2(id,St);
	var s=oo.P=="L"?1:(oo.P=="S"?-1:alert("wtf"));
	var r=0;//0.005;
	var impVol=parseInt($("#impVol").val());
	var d1=calcd1(impVol/100,oo.Tmt,St,oo.K,r);
	var d2=calcd2(impVol/100,oo.Tmt,St,oo.K,r);
	var vc=pnorm(d1)*St-pnorm(d2)*oo.K*Math.pow(Math.E,-r*oo.Tmt); 
	var vp=pnorm(-1*d2)*oo.K*Math.pow(Math.E,-r*oo.Tmt)-pnorm(-1*d1)*St;	//console.log("vc="+vc+",vp="+vp);
	var vs=St-oo.K;
	var v=oo.O=="C"?vc:(oo.O=="P"?vp:(oo.O=="S"?St:alert("wtf2"))); // in case of stock, K is for the purchase price of the stock
	var p=oo.O=="C"||oo.O=="P"?oo.V:0;
	var Tmt=oo.Tmt;
	return(oo.Q*s*(v-p));
}

//
function thalesTheorem_yn(x1,y1,x2,y2,xn) {
	if(y2==y1) { // flat horizontal line
		return y1;
	} else if(x2==x1) { // vertical line
		return y1; // choosing arbitrary point
	} else {
		x0=y1-(y2-y1)/(x2-x1)*x1;
		xy0=-x0*(x2-x1)/(y2-y1);
		return y1*(xn-xy0)/(x1-xy0);
	}
}

//
function searchCosPoints(v,a) {
  // assumes that cosPoints[i][1] are sorted in increasing order
  for(var i=0;i<v.length;i++) {
    if(v[i][0]<=a && v[i+1][0]>=a) {
		//console.log("here",v[i][0]+','+v[i][1]+';'+a+',?;'+v[i+1][0]+','+v[i+1][1]);
		return(thalesTheorem_yn(v[i][0],v[i][1],v[i+1][0],v[i+1][1],a));
	}
  }
  return(0);
}

//
function draw() {
  // calculate strategy P&L

  var St={St:0};
  $('input[type="checkbox"]').each(function(iteration,item) {
    St.St=myJsonParse($("#"+item.id).prop("value")).St;
  }); // as a matter of fact, taking St of any of those would have worked, so no need to iterate
  St=St.St;

  var cosPoints = []; 
  for (var i=0; i<=2*St; i+=0.25){ 
	var sum={x:0};
	$('input[type="checkbox"]').each(function(iteration,item) {
		aa=item.checked?mygv("#"+item.id,i):0;
		sum.x+= aa;
	});
     cosPoints.push([i,sum.x]); 
  } 

  // risk is +- 7 %
  if($("#rd").length != 0) {
    var rd=$("#rd").val()/100;
  } else { var rd=0.07; }
  
  // search for maximum loss between 0.8*St and 1.2*St
  // Go from left to center, then from right to center. This ensures that the minimum closest to the center is chosen
  var cpSt_min_i=0;
  var cpSt_min=9999999999999999; // infinity
  // Go from left to center
  for(var j=0;j<cosPoints.length;j++) {
	var i=cosPoints[j][0];
	var cpSt_x=cosPoints[j][1];
	if(i>=St*(1-rd) && i<=St && cpSt_x<=cpSt_min) {
		cpSt_min=cpSt_x;
		cpSt_min_i=i;
	}
  }
  // Go from right to center
  for(var j=cosPoints.length-1;j>=0;j--) {
	var i=cosPoints[j][0];
	var cpSt_x=cosPoints[j][1];
	if(i>=St && i<=(1+rd)*St && cpSt_x<=cpSt_min) {
		cpSt_min=cpSt_x;
		cpSt_min_i=i;
	}
  }
  
  // plotting and displaying information
  var cpSt    =searchCosPoints(cosPoints,St    );
  var chartPlots = [
	cosPoints,
//	[[cpSt_min_i,cpSt_min],[St,cpSt]],
	//[[cpSt_min_i,cpSt_min],[cpSt_min_i,0]],
	[[St,cpSt]],
	[[cosPoints[0][0],0],[cosPoints[cosPoints.length-1][0],0]],
	[[St,0],[St,cpSt]],
	[[cpSt_min_i,cpSt_min],[cpSt_min_i,0]]
	];
  var risk2=Math.min(0,Math.round(cpSt*100)/100);
  var risk=Math.min(0,Math.round(cpSt_min*100)/100);
 $("#mrStandard").text(risk2);
 $("#mrConservative").text(risk);
  // make plot
  $("#chart3").empty();
  var plot3 = $.jqplot('chart3', chartPlots, {  
      series:[{showMarker:false}],
      title: 'P&L<br> Margin:<br> standard = 100 x '+risk2+'<br> conservative = 100 x '+risk,
      axes:{
        xaxis:{
          label:'S_T',
          labelRenderer: $.jqplot.CanvasAxisLabelRenderer,
          labelOptions: {
            fontFamily: 'Georgia, Serif',
            fontSize: '12pt'
          },
//	      min: 50,
//	      max: 200
        },
        yaxis:{
          labelRenderer: $.jqplot.CanvasAxisLabelRenderer,
          labelOptions: {
            fontFamily: 'Georgia, Serif',
            fontSize: '12pt'
          },
//	      min: -210,
//	      max: +210
        }
	  }, // end axes
      cursor:{
        show: true,
        zoom:true,
        tooltipLocation:'sw'
        }
  });

  // write the margin requirement
//  if( ($("#c1").prop("checked")||$("#p1").prop("checked"))&&!($("#c2").prop("checked")|$("#p2").prop("checked")|$("#s").prop("checked")) ) $("#mr").text("0");
};

//
//var port; //=JSON.parse('[{"P":"L","Q":1,  "K":125,"O":"C","V":3.75,"St":125},{"P":"S","Q":1,  "K":30, "O":"C","V":0.0625,"St":125},{"P":"S","Q":1,  "K":90, "O":"P","V":7,"St":125},{"P":"L","Q":3,        "O":"S","V":0,"St":125}]');
function handlePort(port) {
	if(!$.isArray(port)) port=JSON.parse(port);

	$("#ol").empty();
	$.each(port, function(iteration,item) {
		// update item Tmt to remove the physical time passed
               item.Tmt=item.TMaturity-$('#tPhysical').val();
               // Not doing Math.max to distinguish between expired and at expiry // item.Tmt=Math.max(0,item.TMaturity-$('#tPhysical').val());^M
               // check if expired
               expired=item.Tmt<0;

		// list
		$("#ol").append(
			$(document.createElement("li")
			).attr({
				"style":(expired?"background-color:red":"")
			}).append(
				$(document.createElement("input")).attr({
					"id":'o'+iteration,
					"type":"checkbox",
					"class":"cb",
					"value":JSON.stringify(item),
					"checked":!expired // false if expired
				})
			).append(
				(item.P=="L"?"Long":"Short")
				+" "+item.Q+"x"
				+(item.O=="C"||item.O=="P"?" T-t="+item.Tmt+" years":"")
				+(item.O=="C"||item.O=="P"?" "+item.K:"")
				+(item.O=="C"?" Call":(item.O=="P"?" Put":" Security"))
				+(item.O=="C"||item.O=="P"?" at "+item.V:"")
				+" (underlying at "+item.St+")"
				//+", "+item.Comment+")"
				+", Expired: "+(expired?"Yes":"No")
			)
		);
	});
	$(".cb").on("click", draw);

	$('#impVolText').text($('#impVol').val());
	$('#TMaturityText').text($('#TMaturity').val());
	$('#underlyingText').text($('#St').val());
	$('#rdText' ).text($('#rd').val());
	$('#rdText2').text($('#rd').val());
	$('#tPhysicalText').text($('#tPhysical').val());

	draw();
};

//
function doRetrieve() {
	// inspired by ajaxMgmt.jsa
	reqJson('visOpt/getClientPortfolioForVis.html',
		'mainContent',
		'dd='+document.getElementById('ddFree').value+'&client.base='+document.getElementById('client.base').value+'&cid='+document.getElementById('cid').value+'&secId='+document.getElementById('secId').value+'&format=json',
		false,
		handlePort);
}

// \Todo why doesn't this work here and only works in window.load in ajaxMgmt.js?
//$(document).ready(function() {
//	req('am/list.validAccountIds.html','cid','client.base=Lebanon&format=json');
//alert("bla");
//});
